The CI is still failing with:
```
E           ------------------------ Command failed (unexpectedly): ------------------------
E           install pkg/0.1@ -pr=profile
E           ----------------------------------- Output: ------------------------------------
E           
E           WARN: **************************************************
E           WARN: *** Conan 1 is legacy and on a deprecation path **
E           WARN: *********** Please upgrade to Conan 2 ************
E           WARN: **************************************************
E           Configuration:
E           [settings]
E           arch=x86_64
E           build_type=Release
E           compiler=msvc
E           compiler.cppstd=14
E           compiler.runtime=dynamic
E           compiler.runtime_type=Release
E           compiler.version=191
E           os=Windows
E           [options]
E           [build_requires]
E           [env]
E           
E           Installing package: pkg/0.1
E           Requirements
E               pkg/0.1 from local cache - Cache
E           Packages
E               pkg/0.1:16041e00ee7792b16701a7a92918452f6b5a519b - Missing
E           
E           Cross-build from 'Linux:x86_64' to 'Windows:x86_64'
E           Installing (downloading, building) binaries...
E           ERROR: Missing binary: pkg/0.1:16041e00ee7792b16701a7a92918452f6b5a519b
E           
E           pkg/0.1: WARN: Can't find a 'pkg/0.1' package for the specified settings, options and dependencies:
E           - Settings: arch=x86_64, build_type=Release, compiler=msvc, compiler.cppstd=14, compiler.runtime=dynamic, compiler.runtime_type=Release, compiler.version=191, os=Windows
E           - Options: 
E           - Dependencies: 
E           - Requirements: 
E           - Package ID: 16041e00ee7792b16701a7a92918452f6b5a519b
E           
E           ERROR: Missing prebuilt package for 'pkg/0.1'
E           Use 'conan search pkg/0.1@ --table=table.html -r=remote' and open the table.html file to see available packages
E           Or try to build locally from sources with '--build=pkg'
E           
E           More Info at 'https://docs.conan.io/en/latest/faq/troubleshooting.html#error-missing-prebuilt-package'
```

in ``test_msvc_visual_incompatible`` test

--- a/conans/model/info.py
+++ b/conans/model/info.py
@@ -591,8 +591,14 @@ class ConanInfo(object):
         version = compatible.settings.compiler.version
         runtime = compatible.settings.compiler.runtime
         runtime_type = compatible.settings.compiler.runtime_type
+        cppstd = compatible.settings.compiler.get_safe("cppstd")
 
         compatible.settings.compiler = "Visual Studio"
+
+        if cppstd is not None:
+            # Restore cppstd, as switching to 'Visual Studio' clears it
+            compatible.settings.compiler.cppstd = cppstd
+
         from conan.tools.microsoft.visual import msvc_version_to_vs_ide_version
         visual_version = msvc_version_to_vs_ide_version(version)
         compatible.settings.compiler.version = visual_version
@@ -600,6 +606,13 @@ class ConanInfo(object):
         if runtime_type == "Debug":
             runtime = "{}d".format(runtime)
         compatible.settings.compiler.runtime = runtime
+
+        # Some 'final adjustment'
+        # In case the cppstd is the 'default' for the compiler, it will actually be erased
+        default_cppstd = cppstd_default(compatible.settings)
+        if default_cppstd is not None and compatible.settings.compiler.get_safe("cppstd") == default_cppstd:
+            del compatible.settings.compiler.cppstd
+
         return compatible
 
     def apple_clang_compatible(self):

--- a/conans/test/integration/package_id/compatible_test.py
+++ b/conans/test/integration/package_id/compatible_test.py
@@ -604,7 +604,27 @@ class CompatibleIDsTest(unittest.TestCase):
 def test_msvc_visual_incompatible():
     conanfile = GenConanfile().with_settings("os", "compiler", "build_type", "arch")
     client = TestClient()
-    profile = textwrap.dedent("""
+    # First we need to make sure to test that a 'non-default' cppstd is actually preserved
+    # and we can match that
+    profile1 = textwrap.dedent("""
+        [settings]
+        os=Windows
+        compiler=msvc
+        compiler.version=191
+        compiler.runtime=dynamic
+        compiler.cppstd=17
+        build_type=Release
+        arch=x86_64
+        """)
+    client.save({"conanfile.py": conanfile,
+                 "profile1": profile1})
+    client.run('create . pkg/0.1@ -s os=Windows -s compiler="Visual Studio" -s compiler.version=15 '
+               '-s compiler.cppstd=17 -s compiler.runtime=MD -s build_type=Release -s arch=x86_64')
+    client.run("install pkg/0.1@ -pr=profile1")
+    assert "Using compatible package" in client.out
+
+    # Now we need to test the 'default' cppstd case, in this case, this will be lost in the package id
+    profile2 = textwrap.dedent("""
         [settings]
         os=Windows
         compiler=msvc
@@ -614,15 +634,15 @@ def test_msvc_visual_incompatible():
         build_type=Release
         arch=x86_64
         """)
-    client.save({"conanfile.py": conanfile,
-                 "profile": profile})
-    client.run('create . pkg/0.1@ -s os=Windows -s compiler="Visual Studio" -s compiler.version=15 '
+    client.save({"profile2": profile2})
+    client.run('create . pkg/0.2@ -s os=Windows -s compiler="Visual Studio" -s compiler.version=15 '
                '-s compiler.runtime=MD -s build_type=Release -s arch=x86_64')
-    client.run("install pkg/0.1@ -pr=profile")
+    client.run("install pkg/0.2@ -pr=profile2")
     assert "Using compatible package" in client.out
+
     new_config = "core.package_id:msvc_visual_incompatible=1"
     save(client.cache.new_config_path, new_config)
-    client.run("install pkg/0.1@ -pr=profile", assert_error=True)
+    client.run("install pkg/0.1@ -pr=profile1", assert_error=True)
     assert "ERROR: Missing prebuilt package for 'pkg/0.1'" in client.out
 
 

